lib/srv/usermgmt.go

```diff
@@ -0,0 +1,261 @@
+	// UserCleanup starts a periodic user deletion cleanup loop for
+	// users that failed to delete
+	UserCleanup()
+	// Shutdown cancels the UserCleanup loop
+	Shutdown()
+
+	// UserExists returns nil should a hostuser exist
+	UserExists(string) (*user.User, error)
+}
+
+type HostUserManagement struct {
+	backend HostUsersBackend
+	ctx     context.Context
+	cancel  context.CancelFunc
+	storage *local.PresenceService
+	lease   *types.SemaphoreLease
+}
+
+var _ HostUsers = &HostUserManagement{}
+
+// CreateUser creates a temporary Teleport user in the TeleportServiceGroup
+func (u *HostUserManagement) CreateUser(name string, ui *services.HostUsersInfo) (*user.User, io.Closer, error) {
+	tempUser, err := u.backend.Lookup(name)
+	if err != nil && err != user.UnknownUserError(name) {
+		return nil, nil, trace.Wrap(err)
+	}
+
+	if tempUser != nil {
+		// try to delete even if the user already exists as only users
+		// in the teleport-system group will be deleted and this way
+		// if a user creates multiple sessions the account will
+		// succeed in deletion
+		return tempUser, &userCloser{
+			username: name,
+			users:    u,
+			backend:  u.backend,
+		}, trace.AlreadyExists("User %q already exists", name)
+	}
+
+	groups := append(ui.Groups, types.TeleportServiceGroup)
+	var errs []error
+	for _, group := range groups {
+		if err := u.createGroupIfNotExist(group); err != nil {
+			errs = append(errs, err)
+			continue
+		}
+	}
+	if err := trace.NewAggregate(errs...); err != nil {
+		return nil, nil, trace.WrapWithMessage(err, "error while creating groups")
+	}
+
+	if err := u.storage.UpsertUserLoginTime(u.ctx, name, time.Now()); err != nil {
+		return nil, nil, trace.Wrap(err)
+	}
+
+	err = u.backend.CreateUser(name, groups)
+	if err != nil && !trace.IsAlreadyExists(err) {
+		return nil, nil, trace.WrapWithMessage(err, "error while creating user")
+	}
+
+	tempUser, err = u.backend.Lookup(name)
+	if err != nil {
+		return nil, nil, trace.Wrap(err)
+	}
+
+	return tempUser, &userCloser{
+		username: name,
+		users:    u,
+		backend:  u.backend,
+	}, nil
+}
+
+func (u *HostUserManagement) createGroupIfNotExist(group string) error {
+	_, err := u.backend.LookupGroup(group)
+	if err != nil && err != user.UnknownGroupError(group) {
+		return trace.Wrap(err)
+	}
+	err = u.backend.CreateGroup(group)
+	if trace.IsAlreadyExists(err) {
+		return nil
+	}
+	return trace.Wrap(err)
+}
+
+// DeleteAllUsers deletes all host users in the teleport service group.
+func (u *HostUserManagement) DeleteAllUsers() error {
+	users, err := u.backend.GetAllUsers()
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	teleportGroup, err := u.backend.LookupGroup(types.TeleportServiceGroup)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	var errs []error
+	for _, name := range users {
+		lt, err := u.storage.GetUserLoginTime(u.ctx, name)
+		if err != nil {
+			log.Debugf("Failed to find user %q login time: %s", name, err)
+			continue
+		}
+		if time.Since(lt) < time.Second*30 {
```

Comment:
I think this should be longer than any possible session (and even then, without the setting that closes sessions when the certificate expires this could technically run over the end of a session), but perhaps the RFD should define what's a good time to wait before cleaning up provisioned host users instead?

```suggestion
		if time.Since(lt) < 30*time.Hour {
```

Reply:
Following the decision in https://github.com/gravitational/teleport/pull/11830#discussion_r887819693 I think there should be a comment here saying that the small grace period is just to have enough time to go from the user creation to the shell spawn without having to hold the lock